src/main/starlark/builtins_bzl/common/objc/compilation_support.bzl

```diff
@@ -42,6 +42,31 @@ def _build_variable_extensions(ctx, arc_enabled):
 
     return extensions
 
+def _create_compilation_attributes(ctx):
+    disallow_sdk_frameworks = ctx.fragments.objc.disallow_sdk_frameworks_attributes
+    sdk_frameworks = getattr(ctx.attr, "sdk_frameworks", [])
+    weak_sdk_frameworks = getattr(ctx.attr, "weak_sdk_frameworks", [])
+    if disallow_sdk_frameworks:
+        if sdk_frameworks:
+            fail("sdk_frameworks attribute is disallowed. Use explicit dependencies instead.")
+        if weak_sdk_frameworks:
+            fail("weak_sdk_frameworks attribute is disallowed. Use explicit dependencies instead.")
+
+    return struct(
+        hdrs = depset(cc_helper.get_public_hdrs(ctx)),
+        textual_hdrs = depset(getattr(ctx.files, "textual_hdrs", [])),
+        sdk_includes = depset(getattr(ctx.attr, "sdk_includes", [])),
+        includes = depset(getattr(ctx.attr, "includes", [])),
+        sdk_frameworks = depset(sdk_frameworks),
+        weak_sdk_frameworks = depset(weak_sdk_frameworks),
+        sdk_dylibs = depset(getattr(ctx.attr, "sdk_dylibs", [])),
+        linkopts = objc_internal.expand_and_tokenize(ctx = ctx, attr = "linkopts", flags = getattr(ctx.attr, "linkopts", [])),
```

Comment:
This and data weren't covered by tests for expansions, hoping internal google CI helps validate some of this

Reply:
I'll take a close look what the internal Google tests can check here.

There's one more thing I noticed: in ObjcStarlarkInternal.createCompilationAttributes() the call to CompilationAttributes.Builder.addCompileOptionsFromRuleContext() (which sets linkopts, copts, additional_linker_inputs and defines) is only called if the rule has copts.
After digging a bit into the history of this, I currently think that this was not an intended coupling of conditions; the straightforward linear copying that you implemented makes more sense to me.

We can discuss tweaks if any problems surface from this.